devito/types/grid.py

```diff
@@ -421,3 +425,63 @@ class Interior(SubDomain):
 
     def define(self, dimensions):
         return {d: ('middle', 1, 1) for d in dimensions}
+
+
+class SubDomains(SubDomain):
+    """
+    Class to define a set of N (a positive integer) subdomains.
+
+    Parameters
+    ----------
+    **kwargs
+        * N : int
+            Number of subdomains.
+        * bounds : tuple
+            Tuple of numpy int32 arrays representing the bounds of
+            each subdomain.
+
+    Examples
+    --------
+    Include full example + example of shorthand notation here.
+    """
+
+    def __init__(self, **kwargs):
+        super(SubDomains, self).__init__()
+        self._n_domains = kwargs.get('N', 1)
+        self._bounds = kwargs.get('bounds', None)
+
+    def __subdomain_finalize__(self, dimensions, shape):
+        super(SubDomains, self).__subdomain_finalize__(dimensions, shape)
+        n = Dimension(name='n')
+        self._implicit_dimension = n
+
+    @property
+    def n_domains(self):
+        return self._n_domains
+
+    @property
+    def bounds(self):
+        return self._bounds
+
+    @cached_property
+    def _ie_dat(self):
+        if not len(self._bounds) == 2*len(self.dimensions):
+            raise ValueError("Left and right bounds must be supplied for each dimension")
+        n_domains = self.n_domains
+        i_dim = self._implicit_dimension
+        b_f = {}
+        mMstring = ['_m', '_M']
```

Comment:
you can/should get these strings via 
 https://github.com/opesci/devito/blob/master/devito/types/dimension.py#L132

so maybe fname below is just ``d.min_name`` ? (don't know if I'm interpreting the code correctly)

Reply:
Ta, will check.